plutus-benchmark/bitwise/src/PlutusBenchmark/Ed25519.hs

```diff
@@ -0,0 +1,147 @@
+import PlutusTx.Prelude hiding (inv)
+
+{-# INLINE checkValid #-}
+checkValid ::
+  BuiltinByteString ->
+  BuiltinByteString ->
+  BuiltinByteString ->
+  Bool
+checkValid sig message pubKey =
+  let r = decodePoint (sliceByteString 0 32 sig)
+      a = decodePoint pubKey
+      s = decodeInt (sliceByteString 32 32 sig)
+      hintArg = encodePoint r <> pubKey <> message
+      h = hint hintArg
+   in scalarMult bPoint s == edwards r (scalarMult a h)
+
+-- Helpers
+
+newtype Point = Point (Integer, Integer)
+
+instance Eq Point where
+  {-# INLINE (==) #-}
+  Point (x1, y1) == Point (x2, y2) = x1 == x2 && y1 == y2
+
+{-# INLINE bPoint #-}
+bPoint :: Point
+bPoint = Point (bx `modulo` q, by `modulo` q)
+
+{-# INLINE decodePoint #-}
+decodePoint :: BuiltinByteString -> Point
+decodePoint bs
+  | odd x /= x_0 = Point (q - x, yInt)
+  | otherwise = Point (x, yInt)
+  where
+    x_0 = readBit bs 7
+    x = xRecover yInt
+    yInt = decodeInt $ clearBit 7 bs
+
+{-# INLINE encodePoint #-}
+encodePoint :: Point -> BuiltinByteString
+encodePoint (Point (x, y)) = result
+  where
+    zeroPos :: Integer
+    zeroPos = 7
+    result = writeBits yBS [zeroPos] [xLSBVal]
+    yBS = integerToByteString LittleEndian 32 y
+    xBS = integerToByteString LittleEndian 32 x
+    xLSBVal = readBit xBS 248
+
+{-# INLINE scalarMult #-}
+scalarMult :: Point -> Integer -> Point
+scalarMult p e =
+  if e == 0
+    then Point (0, 1)
+    else
+      let q' = scalarMult p (e `divide` 2)
+          q'' = edwards q' q'
+       in if odd e then edwards q'' p else q''
+
+{-# INLINE edwards #-}
+edwards :: Point -> Point -> Point
+edwards (Point (x1, y1)) (Point (x2, y2)) = Point (x3 `modulo` q, y3 `modulo` q)
+  where
+    x3 :: Integer
+    x3 = (x1 * y2 + x2 * y1) * inv (1 + d * x1 * x2 * y1 * y2)
+    y3 :: Integer
+    y3 = (y1 * y2 + x1 * x2) * inv (1 - d * x1 * x2 * y1 * y2)
+
+{-# INLINE hint #-}
+hint :: BuiltinByteString -> Integer
+hint = byteStringToInteger LittleEndian . sha512
+
+{-# INLINE decodeInt #-}
+decodeInt :: BuiltinByteString -> Integer
+decodeInt = byteStringToInteger LittleEndian
+
+{-# INLINE bx #-}
+bx :: Integer
+bx = xRecover by
+
+{-# INLINE by #-}
+by :: Integer
+by = 4 * inv 5
+
+{-# INLINE q #-}
+-- 2 ^ 255 - 19, but as a constant
+q :: Integer
+q = 57896044618658097711785492504343953926634992332820282019728792003956564819949
+
+{-# INLINE xRecover #-}
+xRecover :: Integer -> Integer
+xRecover y =
+  if
+      | cond1 && not cond2 -> xA
+      | cond1 && cond2     -> xAB
+      | not cond1 && cond2 -> xB
+      | otherwise          -> x
+  where
+    xx :: Integer
+    xx = (y * y - 1) * inv (d * y * y + 1)
+    x :: Integer
+    x = expMod xx ((q + 3) `divide` 8) q
+    xA :: Integer
+    xA = x * i `modulo` q
+    xB :: Integer
+    xB = q - x
+    xAB :: Integer
+    xAB = q - xA
+    cond1 :: Bool
+    cond1 = (x * x - xx) `modulo` q /= 0 -- x here is always the input x
+    cond2 :: Bool
+    cond2 = if cond1 then odd xA else odd x
+
```

Comment:
However, it still seems that the calculation is wrong, although I'm confused by the way that this function interacts with `decodePoint`  For example the values x and q-x are two possible square roots of `xx` and the uses of `even` and `odd` are selecting one of them based on the lowest bit of x, but the selection should in fact be based on the value of one of the bits of the bytestring input to `decodePoint`.  I think that maybe this bit should be fed to `xRecover` and used to select which of x and q-x is returned, and similarly for x' and q-x' (although I see now that `decodePoint` sometimes replaces the x returned by `xRecover` with q-x, so there's some double negation going on which is maybe intended to deal with this:  if so, that's very confusing because the decoding logic is being split over two functions).

In addition, there are elements y of the base field such that there is no x with (x,y) lying on the curve, so given an encoded y-value it may not be possible to find a suitable x, and `xRecover` should fail: this possibility doesn't seem to be accounted for anywhere (I think that `cond` should really be `x * x = -xx modulo q` and `not cond` should be `x * x = xx modulo q` (or vice versa), which leaves cases where neither is true which should lead to failure; furthermore there's a special case when x=0 and the sign bit's set in the encoded bytestring input, which should also fail).

Are you assuming that this code is never applied to an invalid point encoding?  If so, that should be stated in a comment: the code's already difficult to understand and if there are some unstated preconditions that won't help.

Reply:
I took this directly from the reference implementation. Due to its complexity, I was afraid to change it. You're likely right, but I'd need an expert opinion on this to be sure.

Reply:
Thanks.  I'll take a look at the reference implementation.

Reply:
> Thanks. I'll take a look at the reference implementation.

Having looked, I think that what they're doing is returning a potentially invalid x-coordinate from `xRecover` but then later on they have an `isoncurve` check (omitted here) that will reject (x,y) pairs where the x value is wrong.  That still seems a bit suspect, but I should complain to them rather than you.